### PR TITLE
Add support for Nordic proprietary BLE Controller v0.3.0-2.prealpha

### PR DESCRIPTION
**Pull request by kromancer**
Use the new API to determine additional memory requirements per link,
instead of relying on the user to determine `BLECTRL_MEMPOOL_SIZE`

I did a basic smoke-test on the following examples:

sample_name |KConfig changes
----------------|------------------
nrf/samples/bluetooth/throughput|Selected "Nordic proprietary BLE Link Layer"
nrf/samples/bluetooth/central_uart|Selected "Nordic proprietary BLE Link Layer"
nrf/samples/bluetooth/peripheral_uart|Selected "Nordic proprietary BLE Link Layer", changed "BT_HCI_TX_STACK_SIZE" to 1024
zephyr/samples/bluetooth/peripheral_hr|Selected "Nordic proprietary BLE Link Layer", changed "BT_HCI_TX_STACK_SIZE" to 1024

### CODE REVIEW

**Comment by joerchan** [comment]
Added DNM until dependent review in nrfxlib has been merged.

**Comment by jhn-nordic** [comment]
please rebase the PR. Ran into issues with ble HIDS due to the PR outdated.

**Comment by pdunaj** [comment]
does it build? I get
```
In file included from /home/pdunaj/work/ncs/nrf/subsys/bluetooth/controller/hci_driver.c:14:
/home/pdunaj/work/ncs/nrf/subsys/bluetooth/controller/hci_driver.c:42:2: error: 'CONFIG_BT_CTLR_DATA_LENGTH_MAX' undeclared here (not in a function); did you mean 'CONFIG_BT_CTLR_SLAVE_FEAT_REQ'?
  CONFIG_BT_CTLR_DATA_LENGTH_MAX, \
```

File: subsys/bluetooth/controller/hci_driver.c [hunk 1]
```diff
@@ -31,7 +31,21 @@ static K_THREAD_STACK_DEFINE(recv_thread_stack, CONFIG_BLECTLR_RX_STACK_SIZE);
 static K_THREAD_STACK_DEFINE(signal_thread_stack,
 			     CONFIG_BLECTLR_SIGNAL_STACK_SIZE);
 
-static u8_t ble_controller_mempool[CONFIG_BLECTRL_MEMPOOL_SIZE];
+#define MASTER_MEM_SIZE (BLE_CONTROLLER_MEM_PER_MASTER_LINK( \
+	CONFIG_BT_CTLR_DATA_LENGTH_MAX, \
```
Comment (pdunaj) [inline subsys/bluetooth/controller/hci_driver.c]: `CONFIG_BT_CTLR_DATA_LENGTH_MAX` is unavailable if `BT_DATA_LEN_UPDATE` is disabled

**Comment by pdunaj** [comment]
I also see some problems while interacting with bonds
```
[00714663] <inf> ble_bond: Remove peers on identity 0
[00715272] <err> os: ***** HARD FAULT *****
[00715272] <err> os:   Fault escalation (see below)
[00715273] <err> os: ***** MPU FAULT *****
[00715273] <err> os:   Stacking error (context area might be not valid)
[00715273] <err> os:   Data Access Violation
[00715274] <err> os:   MMFAR Address: 0x20009090
[00715274] <err> os: r0/a1:  0x20018ca8  r1/a2:  0x00010046  r2/a3:  0x1ac72444
[00715274] <err> os: r3/a4:  0x2002000c r12/ip:  0x98a5ec32 r14/lr:  0x254296c9
[00715274] <err> os:  xpsr:  0x00010818
[00715274] <err> os: Faulting instruction address (r15/pc): 0x3230845d
[00715275] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception
[00715275] <err> os: Current thread: 0x2000599c (unknown)
[01056289] <err> os: Halting system
```

I increased both ble rx and tx stacks significantly and it seems it work for some time but when I played with bond erasing I got 
```
[03564453] <err> hids_c: PM read error
[03564454] <err> hid_forward: err:14
[03564611] <err> os: ***** HARD FAULT *****
[03564611] <err> os:   Fault escalation (see below)
[03564612] <err> os: ***** MPU FAULT *****
[03564612] <err> os:   Stacking error (context area might be not valid)
[03564612] <err> os:   Data Access Violation
[03564612] <err> os:   MMFAR Address: 0x2000a7d0
[03564613] <err> os: r0/a1:  0x200095bc  r1/a2:  0x00860b39  r2/a3:  0x001772c5
[03564613] <err> os: r3/a4:  0x00000000 r12/ip:  0x01000000 r14/lr:  0x00000000
[03564613] <err> os:  xpsr:  0x61000000
[03564613] <err> os: Faulting instruction address (r15/pc): 0x0001eabc
[03564614] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception
[03564614] <err> os: Current thread: 0x2000599c (unknown)
[03822016] <err> os: Halting system
```

Once dongle crashed on USB but this may be unrelated.
```
[00000082] <inf> event_manager: e: led_event led_id:0 effect:0x4c374
ASSERTION FAIL [state != USB_STATE_DISCONNECTED] @ ../src/modules/usb_state.c:331
[00000165] <err> os: r0/a1:  0x00000004  r1/a2:  0x0000014b  r2/a3:  0x00000000
[00000165] <err> os: r3/a4:  0x0000000c r12/ip:  0x00000000 r14/lr:  0x00026303
[00000166] <err> os:  xpsr:  0x61000000
[00000166] <err> os: Faulting instruction address (r15/pc): 0x00027f24
[00000166] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic
[00000166] <err> os: Current thread: 0x20002498 (unknown)
[00628657] <err> os: Halting system
```

Hi @jhn-nordic , have you seen this issues? Have you tried to switch/erase bonds?

**Comment by pdunaj** [comment]
I think the USB issue is systematic as I get the crash every time I start the dongle with erased memory. Maybe increased delay introduced by the stack (while identities are generated and stored) causes USB to run into a race condition.

**Comment by joerchan** [comment]
> BTW - it is possible that we have not seen this issue so far as development branch where we tested llpm was not rebased against the Zephyr 2.0 where several changes to USB were introduced.

Could we maybe do this is 2 steps, first rebase to Zephyr 2.0, then upgrade the BLE controller version? Then we would know if it is a regression in the BLE controller or not.

**Comment by pdunaj** [comment]
BLE crash is also systematic. It occurs when I initiate bonds erasure on dongle, even if I increase the stack size.

**Comment by pdunaj** [comment]
> > BTW - it is possible that we have not seen this issue so far as development branch where we tested llpm was not rebased against the Zephyr 2.0 where several changes to USB were introduced.
> 
> Could we maybe do this is 2 steps, first rebase to Zephyr 2.0, then upgrade the BLE controller version? Then we would know if it is a regression in the BLE controller or not.

This should be relatively simple, at least for the LLPM part or the dev repo. I will do it and let you know.

**Comment by pdunaj** [comment]
Hi, thanks @rugeGerritsen and Emil.
It seems that for BLE issue we need to increase not only the stack of rx thread but also workqueue stack. This should be no problem at the moment as we have enough RAM.
As for the USB issue the problem is a race and is caused by change in the USB driver introduced by Zephyr 2.0. It is not related but apparently timings change enough with this controller to replicate it.

I will quickly test it and give a tick.

Hi @kromancer , will you rebase change and fix the conflict?

**Comment by jhn-nordic** [comment]
@rugeGerritsen was not able to reproduce the USB issue yesterday with what seems to be the exact same setup and procedure that I use to reproduce it on my desk. I agree that this issue should not block this PR.

**Comment by kromancer** [comment]
@pdunaj Have now rebased. Let us know if you give a "green light" to proceed with:
1. Merging NordicPlayground/nrfxlib#85
2. Updating west.yml and merging this one

**Comment by pdunaj** [comment]
After expanding heaps code is working fine. Increased RAM usage can be something that can be dealt later on, especially that this is a wider problem.
USB issue is not connected to this PR and fix will have to go through Zephyr upstream.

**Comment by pdunaj** [comment]
Hi @kromancer , I think the documentation is failing. I am not sure what is wrong with nrf tests. Maybe it is broken again.

**Comment by kromancer** [comment]
@pdunaj  It seems that warnings in the documentation build are elevated to CI errors.
These doc warnings are also responsible for the red CI/Jenkins/NRF.

The doc warnings are related to `__PACKED_STRUCT` in the ble-controller's headers.
I and @b-gent are on this.

**Comment by pdunaj** [comment]
Hi @kromancer , it's green :) Let's fix the revision in manifest and merge it before CI stops working again.

**Comment by kapi-no** [comment]
Could you CC at least one person from BLE samples maintainers whenever there is a PR that changes the ble_controller modules? 

Here is a list of people that would like to track these changes:
@kapi-no, @grochu, @rakons or @KAGA164

**Comment by pdunaj** [comment]
Hi @kapi-no , I thought we will be the first to use it and no sample will be impacted...

**Comment by kapi-no** [comment]
@pdunaj, you are right - no sample was impacted so far because they all run by default on the Zephyr controller. However, @qung started to test BLE samples with Nordic BLE controller and he ran into some issues that popped up because of controller misconfiguration at the sample level.

I just would like to have more awareness of what is happening around the Nordic BLE controller :) so that we can align the samples accordingly / provide better support for testers.

**Pull request merged**